docs/proposals/pod-resource-management.md

```diff
@@ -0,0 +1,394 @@
+```
+
+#### Best Effort
+
+For pods belonging to the Best effort Qos we don't set any quota.
+
+```
+/BE/cpu.shares = 2  
+/BE/cpu.quota= not set  
+/BE/m.limit_in_bytes = Allocatable - summation of memory requests of all Guaranteed and Burstable pods  
+/BE/PodUID/m.limit = no limit  
+```
+
+Example:
+We have a pod 'Pod5' having Pod Spec given below:
+
+```yaml
+kind: Pod
+metadata:
+    name: Pod5
+spec:
+    containers:
+        name: foo
+            resources:
+        name: bar
+            resources:
+```
+
+Pod5 is classified as Best effort and is hence nested under the BE cgroup
+
+```
+/BE/cpu.shares = 2  
+/BE/cpu.quota= not set  
+/BE/m.limit_in_bytes = Allocatable - 7Gi  
+/BE/Pod5/m.limit = no limit  
+```
+
+### With Systemd
+
+In systemd we have slices for the three top level QoS class. Further each pod is a subslice of exactly one of the three QoS slices. Each container in a pod belongs to a scope nested under the qosclass-pod slice.
+
+Example:  We plan to have the following cgroup hierarchy on systemd systems
+
+```
+/memory/G-PodUID.slice/containerUID.scope
+/cpu,cpuacct/G-PodUID.slice/containerUID.scope
+/memory/Bu.slice/Bu-PodUID.slice/containerUID.scope
+/cpu,cpuacct/Bu.slice/Bu-PodUID.slice/containerUID.scope
+/memory/BE.slice/BE-PodUID.slice/containerUID.scope
+/cpu,cpuacct/BE.slice/BE-PodUID.slice/containerUID.scope
+```
+
+### Hierarchy Outline
+
+- "$Root" is the system root of the node ie. "/" by default and if `--cgroup-root` is specified then the specified cgroup-root is used as "$Root".
+- We have a top level Qos cgroup for the `Bu`(Burstable) and `BE`(Best Effort) Qos classes.
+- But we __dont__ have a separate cgroup for the `G`(Guaranteed) QoS class. Guaranteed pod cgroups are brought up directly under the `Root` cgroup.
+- Each pod has its own cgroup which is nested under the cgroup matching the pod's Qos class.
+- All containers brought up by the pod are nested under the pod's cgroup.
+- system-reserved cgroup contains the system specific processes.
+- kube-reserved cgroup contains the kubelet specific daemons.
+
+```
+$ROOT
+  |
+  +- Pod1
+  |   |
+  |   +- Container1
+  |   +- Container2
+  |   ...
+  +- Pod2
+  |   +- Container3
+  |   ...
+  +- ...
+  |
+  +- Bu
+  |   |
+  |   +- Pod3
+  |   |   |
+  |   |   +- Container4
+  |   |   ...
+  |   +- Pod4
+  |   |   +- Container5
+  |   |   ...
+  |   +- ...
+  |
+  +- BE
+  |   |
+  |   +- Pod5
+  |   |   |
+  |   |   +- Container6
+  |   |   +- Container7
+  |   |   ...
+  |   +- ...
+  |
+  +- System-reserved
+  |   |
+  |   +- system
+  |   +- ...
+  |
+  +- Kube-reserved 
+  |   |
+  |   +- kubelet
+  |   +- docker (optional)
+  |   +- ...
+  |
+```
+
+QoS Policy Design Decisions
+
+- This hierarchy highly prioritizes resource guarantees to the Guaranteed over Burstable and Best Effort Pods.
+- By not having a separate cgroup for the Guaranteed class, the hierarchy allows the Guaranteed pods to burst and utilize all of Node's Allocatable capacity.
```

Comment:
please stop mixing G and Guaranteed, etc.  just be consistent in using the shorthand ;-)